skllm/preprocessing/gpt_vectorizer.py

```diff
@@ -1,39 +1,54 @@
 from typing import Any, List, Optional, Union
+from functools import partial
 
 import numpy as np
 import pandas as pd
 from numpy import ndarray
-from sklearn.base import BaseEstimator as _BaseEstimator
-from sklearn.base import TransformerMixin as _TransformerMixin
+from sklearn.base import BaseEstimator, TransformerMixin
+from joblib import Parallel, delayed
 from tqdm import tqdm
 
-from skllm.openai.embeddings import get_embedding as _get_embedding
-from skllm.openai.mixin import OpenAIMixin as _OAIMixin
-from skllm.utils import to_numpy as _to_numpy
+from skllm.openai.embeddings import get_embedding
+from skllm.openai.mixin import OpenAIMixin
+from skllm.utils import to_numpy
 
 
-class GPTVectorizer(_BaseEstimator, _TransformerMixin, _OAIMixin):
+class GPTVectorizer(BaseEstimator, TransformerMixin, OpenAIMixin):
     def __init__(
         self,
         openai_embedding_model: str = "text-embedding-ada-002",
         openai_key: Optional[str] = None,
         openai_org: Optional[str] = None,
+        batch_size: int = 32,
     ):
         self.openai_embedding_model = openai_embedding_model
         self._set_keys(openai_key, openai_org)
+        self.batch_size = batch_size
 
-    def fit(self, X: Any = None, y: Any = None, **kwargs):
+    def fit(self, X: Any = None, y: Any = None, **kwargs: Any):
         return self
 
     def transform(self, X: Optional[Union[np.ndarray, pd.Series, List[str]]]) -> ndarray:
-        X = _to_numpy(X)
+        X = to_numpy(X)
+        num_samples = len(X)
         embeddings = []
-        for i in tqdm(range(len(X))):
-            embeddings.append(
-                _get_embedding(X[i], self._get_openai_key(), self._get_openai_org())
+
+        def process_texts(texts):
+            return get_embedding(
+                texts,
+                self._get_openai_key(),
+                self._get_openai_org(),
+                model=self.openai_embedding_model,
             )
+
+        with Parallel(n_jobs=-1) as parallel:
+            for i in tqdm(range(0, num_samples, self.batch_size)):
+                batch_texts = X[i: i + self.batch_size]
+                batch_embeddings = parallel(delayed(process_texts)(texts) for texts in batch_texts)
```

Comment:
Isn't this still calling the API for each sample in the batch separately ?

Reply:
the API is called separately for each text in the batch, which may result in multiple API requests. If the goal is to reduce API calls, I can modify the code to make batched API calls instead of individual calls for each text. However, the current implementation processes texts in parallel within each batch, which has better performance.

Reply:
What I am trying to say is:
(A) The `batch_size` parameter, as it is currently used, will be very confusing for the users, given that it has minimal impact on the performance.
(B) Given how huge the overhead of making the API calls is, I think it might be faster to make fewer calls sequentially instead of more calls in parallel. Obviously, in this case, it is also possible to have both: fewer calls and in parallel.

Reply:
Yes that makes sense. Does this version answer what you want?

Reply:
Seems like you are overthinking the problem too much. The solution does not have to be too complicated and your initial commit was almost fine already.

Regarding the last change: 

1) Concatenating the samples together into a single string is definitely not the way to go. Not only this approach produces the output of the wrong shape, but also combines the information from the independent data points. I am not sure why this is even needed, according to OpenAI docs the embedding endpoint natively supports batched inputs. https://platform.openai.com/docs/api-reference/embeddings/create

2) While using thread pool has some advantages, it might result in undesired side effects if used with very large datasets (e.g. hitting the API rate limits, high usage of resources due to unbounded number of workers). Maybe it is better to either stick to joblib as originally planned, or allow the restriction of the number of workers.